Formulate discussion and security considerations sections (#44)
ietf-wg-asdf/instance-information: draft-ietf-asdf-instance-information.md

1. the intended effect of a message, which can either be a report or an update of a Thing's state, and
2. the actual content of the message, which may be freestanding (without a reference to a previous message or state) or relative (with such a reference).

> The distinction of the intended effect into "update" or "state" is comparable to YANG's "config" statement ({{Section 7.21.1 of -yang}}), a flag value indicating whether a definition representation configuration or state data.

Reviewer: Editorial: since this is intended to parallel L233, it should be in the same sequence.

"Update" may not be the best term, as it is also used for "news update" etc., which is a state update [sic].
See also below example...

Maybe "change"?  "modify"?

Author: Good catch! It could be that I actually intended to say “'report' or 'update'” here. However, something like “change” is probably less ambiguous here – we may need to adjust the terminology usage in other places as well, though.

```suggestion
The distinction of the intended effect into "report" or "change" is comparable to YANG's "config" statement ({{Section 7.21.1 of -yang}}), a flag value indicating whether a definition represents configuration or state data.
```

# Discussion

> Instance-related messages close an important semantic gap of SDF, which so far was only capable of describing device classes and their capabilities.

Reviewer: , leaving the exercise of affordances to the ecosystem-specific mechanisms.

Author: Thank you!

```suggestion
Instance-related messages close an important semantic gap of SDF, which so far was only capable of describing device classes and their capabilities, leaving the exercise of affordances to the ecosystem-specific mechanisms.
```

There are, however, a few aspects that need special consideration and have to be addressed by more fine-grained specifications in the future that will buid upon this one.

> Instance-related messages allow for the setting of timestamps that currently relate to the message as a whole.

Reviewer: ```suggestion
Instance-related messages allow for noting down timestamps that currently relate to the message as a whole.
```
(we are not setting time, are we?)

> Since the values contained within an instance-related message might reflect state updates such as measurements that occurred before composing the message, this timestamp cannot accurately reflect the device state at that exact moment.

Reviewer: "state updates" -- see above...

Author: ```suggestion
Since the values contained within an instance-related message might reflect state changes such as measurements that occurred before composing the message, this timestamp cannot accurately reflect the device state at that exact moment.
```

In this sense, instance-related messages really are imperfect "proofshots" when it comes to the state of affordances, also considering that there might be concurrency problems if a device is updated while generating a Snapshot or Delta message.

> With IoT devices, many additional potential time-related problems arise, as devices might have access to (credible) time information {{-raytime}} or can only measure time unreliably due to resource constraints, accuracy problems, and/or their sleeping behavior {{Section 4.4 of -terms}}.

Reviewer: ```suggestion
With IoT devices, many additional potential time-related problems arise, as devices might have access to (credible) time information {{-raytime}} or can only measure time unreliably due to resource constraints, accuracy problems, and/or their sleeping behavior ({{Section 4.4 of -terms}}).
```

In cases like these, timestamps might have to be omitted from instance-related messages or need to be estimated by a more reliable intermediary.

<!-- TODO: these are still open: -->

Discuss expressing a device's "purpose of life" via context information

> Discuss using context information to indicate provence

Reviewer: provenance?

Author: Ahem, yes, good catch :sweat_smile: 

```suggestion
Discuss using context information to indicate provenance
```

I was not entirely sure yet how we could actually indicate provenance with context information. However, I just realized that provenance in the sense that is meant here relates to the device itself (and not to the model or message), right? Would it make sense to include some kind of example for this kind of metadata?

# Security Considerations

The security considerations of {{-sdf}} apply to this document as well.

> Entities processing instance-related messages need to take measures to ensure the provenance of instance-related messages, including authenticity and integraty.

Reviewer: ```suggestion
Entities processing instance-related messages need to take measures to ensure the provenance of instance-related messages, including authenticity and integrity.
```

> For this purpose, they may use the context and state information contained within a given message.

Reviewer: Quite nebulous.

Author: Maybe this sentence could also be removed, as mentioned before, I think I might have misunderstood what provenance was actually referring to.

Instance-related information may might only be available in certain scopes, e.g. when it comes to security-related configuration parameters.

> This information MUST NOT be exposed to unauthorized entities and therefore MUST be excluded from instance-related messages.

Reviewer: Security considerations should generally be deliberative, not normative, unless there is a reason to be different.  How do we provide selective disclosure of information?  should be briefly touch sd-cwt?

Author: I hadn't heard of SD-CWT before, but that looks definitely very interesting relevant for this aspect.

Would a formulation like this one work here?

```suggestion
This information must not be exposed to unauthorized entities and should therefore be excluded from instance-related messages or only selectively disclosed using, for example, Selective Disclosure CBOR Web Tokens (SD-CWT).
```

> Deployments MAY rely on an authorization framework such as ACE-OAuth {{-ace}} to limit the included information or access to messages as a whole to a certain scope.

Reviewer: That cries for an example...
Can/should we do that?

Author: Sure, I will see what I can do here :)
